### PR TITLE
`9FC03F10.c` :OK:

### PR DESCRIPTION
**Pull request by pixel-stuck**

### CODE REVIEW

File: src/9FC03F10.c [hunk 1]
```diff
@@ -1,15 +1,140 @@
...
+void delay(s32 arg0) {
+    s32 i;
 
-INCLUDE_ASM("asm/non_matchings/9FC03F10", initialize_virage_controller);
+    for(i = 0; i < arg0; i++);
+}
 
-INCLUDE_ASM("asm/non_matchings/9FC03F10", initialize_virage_controllers);
+void initialize_virage_controller(void* controller) {
+    s32 temp = (u32)controller & 0xFFFF0000;
```
Comment (Thar0) [inline src/9FC03F10.c]:
```suggestion
    u32 temp = (u32)controller & 0xFFFF0000;
```

File: src/9FC03F10.c [hunk 2]
```diff
@@ -1,15 +1,140 @@
...
+    IO_WRITE(MI_1C_REG, (1000 / temp_s1) + 1);
+    delay(640000 / temp_s1);
+    temp = IO_READ(controller);
+    if (temp & 1) {
+        temp_lo = 20000 / temp_s1;
+        delay(temp_lo);
+        delay(temp_lo);
+        temp = IO_READ(controller);
+        if (temp & 1) {
+	        return -1;
```
Comment (Thar0) [inline src/9FC03F10.c]:
```suggestion
            return -1;
```
Similar elsewhere (tabs used for indentation)

File: src/9FC03F10.c [hunk 3]
```diff
@@ -1,15 +1,140 @@
...
+        }
+    }
+
+    temp = IO_READ(controller);
+    if ((temp & 0x40000000) == 0) {
```
Comment (Thar0) [inline src/9FC03F10.c]:
```suggestion
    if (!(temp & 0x40000000)) {
```
Very minor but I think these kinds of checks are a bit nicer like this

File: src/9FC03F10.c [hunk 4]
```diff
@@ -1,15 +1,140 @@
...
+    IO_WRITE(temp, 0x02000000);
+    delay(100);
+    var_v1 = IO_READ(temp2);
+    if ((var_v1 & 0x40000000) != 0) {
+        return -1;
+    }
+
+    for(var_v1 = IO_READ(temp2); (var_v1 & 0x40000000) == 0; var_v1 = IO_READ(temp2)) {
+        delay(100);
+    }
```
Comment (Thar0) [inline src/9FC03F10.c]: The duplicated assignment for `var_v1` makes it a bit weird to read, does something like the following work?
```c
while (!(IO_READ(temp2) & 0x40000000)) {
    delay(100);
}
```

**Review by Thar0 — COMMENTED** [review]
Just a couple of small nits, otherwise good to go

File: src/9FC03F10.c (hunk 4 above)
Comment (pixel-stuck) [reply src/9FC03F10.c]: unfortunately not. It seems the devs had a habit of using temps for the IO_READs

**Review by Thar0 — APPROVED** [review]

**Pull request merged**